storage/engine: add Engine.{WriteBatch,BatchRepr} methods

petermattis opened the pull request:
`Engine.BatchRepr` can be used to retrieve the underlying representation
of a batch so that it can be transferred to a remote
node. `Engine.WriteBatch` can then be used to apply the batch. Doing so
is equivalent to calling `Engine.Commit` on the original batch.

Fixes #6288.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6629)

<!-- Reviewable:end -->

petermattis committed:
storage/engine: add Engine.{ApplyBatchRepr,BatchRepr} methods

`Engine.BatchRepr` can be used to retrieve the underlying representation
of a batch so that it can be transferred to a remote
node. `Engine.ApplyBatchRepr` can then be used to apply the batch. Doing
so is equivalent to calling `Engine.Commit` on the original batch.

Fixes #6288.

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> #### storage/engine: add Engine\.\{WriteBatch,BatchRepr\} methods
> 
> `Engine.BatchRepr` can be used to retrieve the underlying representation
> of a batch so that it can be transferred to a remote
> node. `Engine.WriteBatch` can then be used to apply the batch. Doing so
> is equivalent to calling `Engine.Commit` on the original batch.
> 
> Fixes #6288.
> 
> 
> </details>


---

Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[storage/engine/rocksdb/db.h, line 109 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6629#-KHVcd5l5qjA2dS__t5W:-KHVcd5l5qjA2dS__t5X:1814224182) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b3e37c8830e5062d18b941dffdbf76faf690295/storage/engine/rocksdb/db.h#L109)):*
> ```objc
> // WriteBatch writes a batch of mutations encoded using that batch
> // representation returned by DBBatchRepr().
> DBStatus DBWriteBatch(DBEngine* db, DBSlice repr);
> ```

Rename to `DBApplyBatchRepr`? (and similar for the go methods) And unlike the surrounding methods, this one is called on a regular DBEngine, not a batch.

---

*[storage/engine/rocksdb/db.h, line 111 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6629#-KHVcUo3wUx8ARcyuW3M:-KHVcUo3wUx8ARcyuW3N:-1879544348) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b3e37c8830e5062d18b941dffdbf76faf690295/storage/engine/rocksdb/db.h#L111)):*
> ```objc
> DBStatus DBWriteBatch(DBEngine* db, DBSlice repr);
> 
> // Returns the internal batch representation. The returned value is
> ```

Comment that this is only valid on engines created by DBNewBatch.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6629#-:-KHVjvzMF1sHHDqTQCb7:1417088721)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> :lgtm:</details>


---

Reviewed 3 of 5 files at r1.
Review status: 3 of 5 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[storage/engine/engine.go, line 132 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6629#-KHVe9z2BUXLVbQ5FggC:-KHVe9z2BUXLVbQ5FggD:-1883528012) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b3e37c8830e5062d18b941dffdbf76faf690295/storage/engine/engine.go#L132)):*
> ```Go
> 	// The logic for merges is written in db.cc in order to be compatible with RocksDB.
> 	Merge(key MVCCKey, value []byte) error
> 	// WriteBatch atomically applies a set of batched updates. Created by calling
> ```

nit: Why not pull this method down next to its related method?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6629#-:-KHVlZsCqAJkpkxSQFgH:1417088721)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 5 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[storage/engine/engine.go, line 132 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6629#-KHVe9z2BUXLVbQ5FggC:-KHVm4p-Fuv3Pl3bs352:2118351516) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b3e37c8830e5062d18b941dffdbf76faf690295/storage/engine/engine.go#L132)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> nit: Why not pull this method down next to its related method?</details>
No reason. Done.

---

*[storage/engine/rocksdb/db.h, line 109 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6629#-KHVcd5l5qjA2dS__t5W:-KHVkxVHxtL0CycGxcLP:1813969684) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b3e37c8830e5062d18b941dffdbf76faf690295/storage/engine/rocksdb/db.h#L109)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Rename to `DBApplyBatchRepr`? (and similar for the go methods) And unlike the surrounding methods, this one is called on a regular DBEngine, not a batch.</details>
Yeah, I like `ApplyBatchRepr` better. Renamed. Added a comment this doesn't work on batches and a TODO in the implementation that we could add support for batches if needed.

---

*[storage/engine/rocksdb/db.h, line 111 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6629#-KHVcUo3wUx8ARcyuW3M:-KHVlDmOOneHT-znpcTG:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b3e37c8830e5062d18b941dffdbf76faf690295/storage/engine/rocksdb/db.h#L111)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Comment that this is only valid on engines created by DBNewBatch.</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6629)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> :lgtm:</details>


---

Reviewed 5 of 5 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[storage/engine/rocksdb.go, line 224 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6629#-KHVnQBZQ8GhOBSyQFMp:-KHVnQBZQ8GhOBSyQFMq:-846678721) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b3e37c8830e5062d18b941dffdbf76faf690295/storage/engine/rocksdb.go#L224)):*
> ```Go
> 
> // WriteBatch atomically applies a set of batched updates. Created by calling
> // Repr() on a batch. Using this method is equivalent to calling Commit() on a
> ```

Was mildly confused by the comment. Maybe (here and elsewhere)

> Using this method is equivalent to constructing and committing a batch whose Repr() equals repr.

---

*[storage/engine/rocksdb.go, line 500 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6629#-KHVnumKCxTPgDRuOODT:-KHVnumL4Dnem6O2I9Vb:-1414414037) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b3e37c8830e5062d18b941dffdbf76faf690295/storage/engine/rocksdb.go#L500)):*
> ```Go
> }
> 
> // Repr is not implemented for rocksDBSnapshot.
> ```

Seems like we're getting close enough for splitting up that one-size-fits-all interface. It's assembled enough oddities (not suggesting doing it in this PR).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6629#-:-KHVocvUMeIF9siBTsFC:1417088721)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[storage/engine/rocksdb.go, line 224 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6629#-KHVnQBZQ8GhOBSyQFMp:-KHWu_ePKsGbvYdNQl6q:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b3e37c8830e5062d18b941dffdbf76faf690295/storage/engine/rocksdb.go#L224)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Was mildly confused by the comment. Maybe (here and elsewhere)
> 
> > Using this method is equivalent to constructing and committing a batch whose Repr() equals repr.</details>
Done.

---

*[storage/engine/rocksdb.go, line 500 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6629#-KHVnumKCxTPgDRuOODT:-KHWuat61v3hr-RLRYqr:-1748633429) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b3e37c8830e5062d18b941dffdbf76faf690295/storage/engine/rocksdb.go#L500)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Seems like we're getting close enough for splitting up that one-size-fits-all interface. It's assembled enough oddities (not suggesting doing it in this PR).</details>
Yeah, it is slightly awkward to fit everything into the `Engine` interface. Might be worthwhile to have a separate `ReadOnlyEngine` interface which `Engine` embeds.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6629)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6629)*
<!-- Sent from Reviewable.io -->
